datanode/service/party_stats.go

```diff
@@ -159,51 +161,53 @@ func (s *PSvc) GetPartyStats(ctx context.Context, partyID string, markets []stri
 	if err != nil {
 		return nil, err
 	}
+	var rebate num.Decimal
 	if len(vrStats) > 0 {
 		tier, err := s.getVolumeRebateTier(ctx, vrStats[0])
 		if err != nil {
 			return nil, err
 		}
-		rebate, err := num.DecimalFromString(vrStats[0].AdditionalRebate)
+		rebate, err = num.DecimalFromString(vrStats[0].AdditionalRebate)
 		if err != nil {
 			return nil, err
 		}
-		pfFactors.rebate = rebate
 		if tier != nil {
 			data.VolumeRebateTier = *tier.TierNumber
 		}
 	}
 	for _, mkt := range mkts {
 		// @TODO ensure non-nil slice!
-		if err := setMarketFees(data, mkt, pfFactors); err != nil {
+		if err := setMarketFees(data, mkt, rfDiscountFactors, rfRewardFactors, vdDiscountFactors, rebate); err != nil {
 			return nil, err
 		}
 	}
 	return data, nil
 }
 
-func setMarketFees(data *v2.GetPartyDiscountStatsResponse, mkt entities.Market, factors partyFeeFactors) error {
-	maker, infra, liquidity, err := feeFactors(mkt)
+func setMarketFees(data *v2.GetPartyDiscountStatsResponse, mkt entities.Market, rfDiscount, rfRewards, vdFactors partyFeeFactors, rebate num.Decimal) error {
+	maker, infra, liquidity, bb, treasury, err := feeFactors(mkt)
 	if err != nil {
 		return err
 	}
 	// undiscounted
-	base := num.DecimalZero().Add(maker).Add(infra).Add(liquidity)
+	base := num.DecimalZero().Add(maker).Add(infra).Add(liquidity).Add(bb).Add(treasury)
 	// discounted
-	discounted := num.DecimalZero().Add(maker.Sub(factors.maker)).
-		Add(infra.Sub(factors.infra)).
-		Add(liquidity.Sub(factors.liquidity))
+	discountedMaker := maker.Mul(num.DecimalOne().Sub(rfDiscount.maker)).Mul(num.DecimalOne().Sub(vdFactors.maker)).Mul(num.DecimalOne().Sub(rfRewards.maker))
+	discountedInfra := infra.Mul(num.DecimalOne().Sub(rfDiscount.infra)).Mul(num.DecimalOne().Sub(vdFactors.infra)).Mul(num.DecimalOne().Sub(rfRewards.infra))
+	discountedLiquidity := liquidity.Mul(num.DecimalOne().Sub(rfDiscount.liquidity)).Mul(num.DecimalOne().Sub(vdFactors.liquidity)).Mul(num.DecimalOne().Sub(rfRewards.liquidity))
+	discounted := discountedMaker.Add(discountedInfra).Add(discountedLiquidity).Add(bb).Add(treasury)
+
 	data.PartyMarketFees = append(data.PartyMarketFees, &v2.MarketFees{
 		MarketId:             mkt.ID.String(),
 		UndiscountedTakerFee: base.String(),
 		DiscountedTakerFee:   discounted.String(),
 		BaseMakerRebate:      maker.String(),
-		UserMakerRebate:      maker.Add(factors.rebate).String(),
+		UserMakerRebate:      maker.Add(rebate).String(),
```

Comment:
Could we apply the rebate cap here or elsewhere please (buyback + treasury) so the API returns the final rebate factor applied to trades?

Reply:
done